File: app/src/main/java/com/ghstudios/android/features/quests/QuestDetailFragment.java
package com.ghstudios.android.features.quests;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ghstudios.android.AppSettings;
import com.ghstudios.android.AssetLoader;
import com.ghstudios.android.components.ColumnLabelTextCell;
import com.ghstudios.android.components.TitleBarCell;
import com.ghstudios.android.data.classes.Location;
import com.ghstudios.android.data.classes.Quest;
import com.ghstudios.android.data.database.DataManager;
import com.ghstudios.android.mhgendatabase.R;
import com.ghstudios.android.features.locations.LocationDetailPagerActivity;

import java.io.IOException;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.ghstudios.android.mhgendatabase.R.id.location;

public class QuestDetailFragment extends Fragment {
    private static final String ARG_QUEST_ID = "QUEST_ID";
    
    private Quest mQuest;
    private View mView;

    @BindView(R.id.location_layout) LinearLayout mQuestLocationLayout;
    @BindView(R.id.location_image) ImageView questLocationImageView;

    @BindView(R.id.titlebar) TitleBarCell titleBarCell;
    @BindView(R.id.level) TextView levelTextView;
    @BindView(R.id.hub) TextView hubTextView;
    @BindView(R.id.reward) ColumnLabelTextCell rewardCell;
    @BindView(R.id.hrp) ColumnLabelTextCell hrpCell;
    @BindView(R.id.fee) ColumnLabelTextCell feeCell;

    @BindView(R.id.goal) TextView goalTextView;
    @BindView(R.id.location) TextView locationTextView;
    @BindView(R.id.subquest) TextView subquestTextView;
    @BindView(R.id.subhrp) TextView subquestHrpTextView;
    @BindView(R.id.subreward) TextView subRewardTextView;
    @BindView(R.id.description) TextView mFlavor;

    public static QuestDetailFragment newInstance(long questId) {
        Bundle args = new Bundle();
        args.putLong(ARG_QUEST_ID, questId);
        QuestDetailFragment f = new QuestDetailFragment();
        f.setArguments(args);
        return f;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        QuestDetailViewModel viewModel = ViewModelProviders.of(getActivity()).get(QuestDetailViewModel.class);

        viewModel.getQuest().observe(this, quest -> {
            mQuest = quest;
            updateUI();
        });

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_quest_detail, container, false);

        ButterKnife.bind(this, view);

        // Click listener for quest location
        mQuestLocationLayout.setOnClickListener(v -> {
            // The id argument will be the Monster ID; CursorAdapter gives us this
            // for free
            Intent i = new Intent(getActivity(), LocationDetailPagerActivity.class);
            long id = (long)v.getTag();
            if(id>100) id = id-100;
            i.putExtra(LocationDetailPagerActivity.EXTRA_LOCATION_ID, id);
            startActivity(i);
        });

        return view;
    }

    
    private void updateUI() {
        // Add list of monsters and habitats
        FragmentManager fm = getFragmentManager();
        fm.beginTransaction().add(
                R.id.monster_habitat_fragment,
                QuestMonsterFragment.newInstance(mQuest.getId())
        ).commitAllowingStateLoss();

        String cellGoal = mQuest.getGoal();
        String cellHrp = "" + mQuest.getHrp();
        String cellReward = "" + mQuest.getReward() + "z";
        String cellFee = "" + mQuest.getFee() + "z";
        String cellLocation = mQuest.getLocation().getName();
        String cellSubGoal = mQuest.getSubGoal();
        String cellSubHrp = "" + mQuest.getSubHrp();
        String cellSubReward = "" + mQuest.getSubReward() + "z";
        String flavor = mQuest.getFlavor();

        // bind title bar
        titleBarCell.setIconDrawable(AssetLoader.loadIconFor(mQuest));
        titleBarCell.setTitleText(mQuest.getName());

        // bind details section
        hubTextView.setText(AssetLoader.localizeHub(mQuest.getHub()));
        levelTextView.setText(mQuest.getStarString());
        hrpCell.setValueText(cellHrp);
        rewardCell.setValueText(cellReward);
        feeCell.setValueText(cellFee);

        goalTextView.setText(cellGoal);
        locationTextView.setText(cellLocation);
        locationTextView.setTag(mQuest.getLocation().getId());
        subquestTextView.setText(cellSubGoal);
        subquestHrpTextView.setText(cellSubHrp);
        subRewardTextView.setText(cellSubReward);
        mQuestLocationLayout.setTag(mQuest.getLocation().getId());
        mFlavor.setText(flavor);

        // Get Location based on ID and set image thumbnail
        AssetLoader.setIcon(questLocationImageView, mQuest.getLocation());
    }
}

File: app/src/main/java/com/ghstudios/android/features/quests/QuestDetailFragment.kt
package com.ghstudios.android.features.quests

import androidx.lifecycle.Observer
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.lifecycle.ViewModelProvider

import com.ghstudios.android.AssetLoader
import com.ghstudios.android.data.classes.Quest
import com.ghstudios.android.mhgendatabase.R
import com.ghstudios.android.features.locations.LocationDetailPagerActivity

import com.ghstudios.android.ClickListeners.MonsterClickListener
import com.ghstudios.android.data.classes.MonsterToQuest
import com.ghstudios.android.mhgendatabase.databinding.FragmentQuestDetailBinding
import com.ghstudios.android.util.applyArguments
import com.ghstudios.android.util.setImageAsset

/**
 * Shows the main quest information.
 * This class's views are binded via Kotlin KTX extensions
 */
class QuestDetailFragment : Fragment() {

    companion object {
        private const val ARG_QUEST_ID = "QUEST_ID"

        @JvmStatic fun newInstance(questId: Long): QuestDetailFragment {
            return QuestDetailFragment().applyArguments {
                putLong(ARG_QUEST_ID, questId)
            }
        }
    }

    private lateinit var binding: FragmentQuestDetailBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentQuestDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val viewModel = ViewModelProvider(activity!!).get(QuestDetailViewModel::class.java)

        viewModel.quest.observe(viewLifecycleOwner, Observer { quest ->
            if (quest != null) updateUI(quest)
        })

        viewModel.monsters.observe(viewLifecycleOwner, Observer { monsters ->
            if (monsters != null) bindMonsters(monsters)
        })

        // Click listener for quest location
        binding.locationLayout.setOnClickListener { v ->
            // The id argument will be the Monster ID; CursorAdapter gives us this
            // for free
            val i = Intent(activity, LocationDetailPagerActivity::class.java)
            var id = v.tag as Long
            if (id > 100) id = id - 100
            i.putExtra(LocationDetailPagerActivity.EXTRA_LOCATION_ID, id)
            startActivity(i)
        }
    }

    /**
     * Renders the list of monsters in the Quest Detail.
     * Internally, it creates an adapter and then uses it to populate a linear layout.
     */
    private fun bindMonsters(monsters: List<MonsterToQuest>) {
        // Use adapter to manually populate a LinearLayout
        val adapter = MonsterToQuestListAdapter(context!!, monsters)
        for (i in 0 until adapter.count) {
            val v = adapter.getView(i, null, binding.monsterHabitatFragment)
            binding.monsterHabitatFragment.addView(v)
        }
    }

    private fun updateUI(mQuest: Quest) {
        // bind title bar
        with (binding.titlebar) {
            setIconDrawable(AssetLoader.loadIconFor(mQuest))
            setTitleText(mQuest.name)
        }

        with(binding) {
            goal.text = mQuest.goal

            hub.text = AssetLoader.localizeHub(mQuest.hub)
            level.text = mQuest.starString
            hrp.setValueText(mQuest.hrp.toString())
            reward.setValueText("" + mQuest.reward + "z")
            fee.setValueText("" + mQuest.fee + "z")

            location.text = mQuest.location?.name
            location.tag = mQuest.location?.id
            locationLayout.tag = mQuest.location?.id
            subquest.text = mQuest.subGoal
            subhrp.text = "" + mQuest.subHrp
            subreward.text = "" + mQuest.subReward + "z"
            description.text = mQuest.flavor

            // Get Location based on ID and set image thumbnail
            locationImage.setImageAsset(mQuest.location)
        }
    }

    /**
     * Internal adapter used to render the list of monsters shown in a quest
     */
    private class MonsterToQuestListAdapter(context: Context,
                                            items: List<MonsterToQuest>) : ArrayAdapter<MonsterToQuest>(context, 0, items) {

        override fun getView(position: Int, view: View?, parent: ViewGroup): View {
            var view = view
            if (view == null) {
                val inflater = LayoutInflater.from(context)
                view = inflater.inflate(R.layout.fragment_quest_monstertoquest, parent, false)
            }
            val monsterToQuest = getItem(position)

            // Set up the text view
            val itemLayout = view!!.findViewById<LinearLayout>(R.id.listitem)
            val habitatLayout = view.findViewById<LinearLayout>(R.id.habitat_layout)
            val monsterImageView = view.findViewById<ImageView>(R.id.detail_monster_image)
            val monsterTextView = view.findViewById<TextView>(R.id.detail_monster_label)
            val unstableTextView = view.findViewById<TextView>(R.id.detail_monster_unstable)
            val startTextView = view.findViewById<TextView>(R.id.habitat_start)
            val travelTextView = view.findViewById<TextView>(R.id.habitat_travel)
            val endTextView = view.findViewById<TextView>(R.id.habitat_end)
            val hyperTextView = view.findViewById<TextView>(R.id.detail_monster_hyper)

            val cellMonsterText = monsterToQuest!!.monster!!.name

            if (monsterToQuest.isUnstable) {
                unstableTextView.visibility = View.VISIBLE
                unstableTextView.setText(R.string.unstable)
            } else
                unstableTextView.visibility = View.GONE

            if (monsterToQuest.isHyper) {
                hyperTextView.visibility = View.VISIBLE
                hyperTextView.setText(R.string.hyper)
            } else
                hyperTextView.visibility = View.GONE

            monsterTextView.text = cellMonsterText

            AssetLoader.setIcon(monsterImageView, monsterToQuest.monster!!)

            val habitat = monsterToQuest.habitat
            if (habitat != null) {
                val start = habitat.start
                val area = habitat.areas
                val rest = habitat.rest

                var areas = ""
                for (j in area!!.indices) {
                    areas += java.lang.Long.toString(area[j])
                    if (j != area.size - 1) {
                        areas += ", "
                    }
                }

                startTextView.text = java.lang.Long.toString(start)
                travelTextView.text = areas
                endTextView.text = java.lang.Long.toString(rest)
                habitatLayout.visibility = View.VISIBLE
            } else
                habitatLayout.visibility = View.GONE

            itemLayout.tag = monsterToQuest.monster!!.id
            itemLayout.setOnClickListener(MonsterClickListener(context,
                    monsterToQuest.monster!!.id))
            return view
        }
    }
}
